src/LsmTree.hs

```diff
@@ -26,9 +28,12 @@ data Tree a
 instance Foldable Tree where
     foldMap _ Leaf = mempty
     foldMap f (TreeNode l x r) = foldMap f l <> f x <> foldMap f r
+    -- ^ preorder traverse by default
+    foldr _ ini Leaf = ini
+    foldr f ini (TreeNode l x r) = f x (foldr f (foldr f ini r) l)
 
 sizeTree :: Tree a -> Int
-sizeTree = foldr (const (+1)) 0
+sizeTree = foldl' ((succ .) . const) 0
```

Comment:
Never write such code :slightly_smiling_face: It's obscure and hard to read. The following reads better:

```haskell
sizeTree = foldl' (\len _ -> len + 1) 0
```